server/service/src/requisition/program_settings/mod.rs

```diff
@@ -0,0 +1,322 @@
+mod map;
+mod prepare;
+
+use map::map;
+use prepare::prepare;
+
+use repository::{
+    PeriodRow, ProgramRequisitionOrderTypeRow, ProgramRequisitionSettings, ProgramSupplier,
+    RepositoryError,
+};
+
+use crate::service_provider::ServiceContext;
+
+#[derive(Debug, PartialEq)]
+pub struct OrderType {
+    pub order_type: ProgramRequisitionOrderTypeRow,
+    pub available_periods: Vec<PeriodRow>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ProgramSettings {
+    pub program_requisition_settings: ProgramRequisitionSettings,
+    pub suppliers: Vec<ProgramSupplier>,
+    pub order_types: Vec<OrderType>,
+}
+
+// TODO only return and calculate for X number of periods from now ? (back and forward)
+
+/// Method will calculate program settings for a store, broken down into two tasks, prepare and map
+/// See prepare and map for detailed descriptions
+pub(super) fn get_program_requisition_settings(
+    ctx: &ServiceContext,
+    store_id: &str,
+) -> Result<Vec<ProgramSettings>, RepositoryError> {
+    // Get program_settings, order_types, periods and requisitions_in_periods for a store
+    let Some(prepared) =
+        prepare(ctx, store_id)? else {
+            return Ok(Vec::new())
+        };
+
+    // Map program_settings, order_types, periods and requisitions_in_periods to ProgramSettings
+    Ok(map(prepared))
+}
+
+#[cfg(test)]
+mod test {
+    use repository::{
+        mock::{
+            mock_name_store_a, mock_name_store_b, mock_name_store_c, mock_store_a, mock_store_b,
+            mock_store_c, MockData, MockDataInserts,
+        },
+        MasterListNameJoinRow, MasterListRow, Name, NameStoreJoinRow, NameTagJoinRow, NameTagRow,
+        PeriodRow, PeriodScheduleRow, ProgramRequisitionOrderTypeRow, ProgramRequisitionSettings,
+        ProgramRequisitionSettingsRow, ProgramRow, ProgramSupplier, RequisitionRow,
+    };
+
+    use crate::{
+        requisition::program_settings::{OrderType, ProgramSettings},
+        test_helpers::{setup_all_with_data_and_service_provider, ServiceTestContext},
+    };
+
+    #[actix_rt::test]
+    async fn get_program_requisition_settings() {
+        // Mock Data
```

Comment:
I can see you point about setting up mock data!

I'm not 100% sure I follow all of the detail in the tests but looks good to me.

One thing I've started doing in test cases like this, is mutating the database state and then making new assertions.
Example Here:https://github.com/openmsupply/health-supply-hub/blob/5f33702039acb05ba1add04f5b18d865e1dc6dd0/backend/service/src/quote/tests/query_quote_lines.rs#L97
then
https://github.com/openmsupply/health-supply-hub/blob/5f33702039acb05ba1add04f5b18d865e1dc6dd0/backend/service/src/quote/tests/query_quote_lines.rs#L114

In this context I think it might be a bit easier to follow.
E.g.
1. Check available periods for Program A, Order type X
2. Assert - All Periods should be available
3. Add a requisition to Period X 
4. Assert Period X is no longer available (as max_requistions per Period is 1)

I find this `story` style test easier to follow and reason about.
Just an idea to consider.

Reply:
Thanks, for the suggestions

> One thing I've started doing in test cases like this, is mutating the database state and then making new assertions.

That's a good idea, i've used it in follow up tests, probably should have used it more. For this particular test, as it was done last, i was only really testing the mapping and that everything works, other tests are a bit more involved

https://github.com/openmsupply/open-msupply/blob/c110534053a11d5652fba9780679ceb0064421bc/server/repository/src/db_diesel/program_requisition/program_supplier.rs#L257

> I find this story style test easier to follow and reason about.

Oki, will take note, so start with the smallest possible result, and build on from it ? (and maybe the comments can indicate the configuration changes based on users steps ?)